source.go

```diff
@@ -229,177 +207,96 @@ func (s *Source) populateIDs(ctx context.Context) error {
-	err := json.Unmarshal(sdkPos, &pos)
+func (s *Source) getPayload(ctx context.Context, pg client.Page) (sdk.RawData, error) {
+	plainText, err := pg.PlainText(ctx)
 	if err != nil {
-		return position{}, fmt.Errorf("failed unmarshalling position: %w", err)
-	}
-	return pos, nil
-}
-
-func (s *Source) getPayload(
-	ctx context.Context,
-	children notion.Blocks,
-	metadata map[string]string,
-) (sdk.RawData, error) {
-	var plainText string
-	for _, c := range children {
-		text, err := extractText(c)
-		if errors.Is(err, errNoExtractor) {
-			sdk.Logger(ctx).Warn().
-				Str("block_type", c.GetType().String()).
-				Msg("no text extractor registered")
-			continue
-		}
-		if err != nil {
-			return nil, err
-		}
-		plainText += text + "\n"
+		return nil, err
 	}
-
 	payload := recordPayload{
 		Plaintext: plainText,
-		Metadata:  metadata,
+		Metadata:  s.getMetadata(pg),
 	}
 	return json.Marshal(payload)
 }
 
-func (s *Source) getMetadata(page *notion.Page) map[string]string {
+func (s *Source) getMetadata(pg client.Page) map[string]string {
 	return map[string]string{
-		"notion.title":          s.getPageTitle(page),
-		"notion.url":            page.URL,
-		"notion.createdTime":    page.CreatedTime.Format(time.RFC3339),
-		"notion.lastEditedTime": page.LastEditedTime.Format(time.RFC3339),
-		"notion.createdBy":      s.toJSON(page.CreatedBy),
-		"notion.lastEditedBy":   s.toJSON(page.LastEditedBy),
-		"notion.archived":       strconv.FormatBool(page.Archived),
-		"notion.parent":         s.toJSON(page.Parent),
+		"notion.title":          pg.Title(),
+		"notion.url":            pg.URL,
+		"notion.createdTime":    pg.CreatedTime.Format(time.RFC3339),
+		"notion.lastEditedTime": pg.LastEditedTime.Format(time.RFC3339),
+		"notion.createdBy":      pg.CreatedBy,
+		"notion.lastEditedBy":   pg.LastEditedBy,
+		"notion.archived":       strconv.FormatBool(pg.Archived),
+		"notion.parent":         pg.Parent,
 	}
 }
 
-// toJSON converts `v` into a JSON string.
-// In case that's not possible, the function returns an empty string.
-func (s *Source) toJSON(v any) string {
-	bytes, err := json.Marshal(v)
-	if err != nil {
-		return ""
-	}
-	return string(bytes)
-}
-
-// getPageTitle returns the input page's title.
-// In case that's not possible, the function returns an empty string.
-func (s *Source) getPageTitle(page *notion.Page) string {
-	if page == nil || len(page.Properties) == 0 {
-		return ""
-	}
-
-	tp, ok := page.Properties["title"].(*notion.TitleProperty)
-	if !ok || len(tp.Title) == 0 {
-		return ""
-	}
-
-	return tp.Title[0].PlainText
-}
-
-func (s *Source) notFound(err error) bool {
-	nErr, ok := err.(*notion.Error)
-	if !ok {
-		return false
-	}
-	return nErr.Status == http.StatusNotFound
-}
-
 // savePosition saves the position, if it's safe to do so.
 func (s *Source) savePosition(t time.Time) {
+	// see discussion in docs/cdc.md
+	lastTopMinute := time.Now().Truncate(time.Minute)
+	if t.After(lastTopMinute) {
+		return
+	}
```

Comment:
This part was in the search method, so I moved it here since it's more relevant for saving positions. The logic needs to be refactored a bit, to make it more clear. I'll do that separately.